sample/normal.go

```diff
@@ -19,60 +19,136 @@ package sample
 import (
 	"math"
 	"math/big"
-	"math/rand"
 )
 
 // Normal samples random values from the Normal (Gaussian)
 // probability distribution, centered on 0.
 type Normal struct {
-	// Mean
-	// mu float64 // TODO Add if necessary. Currently mu = 0 is assumed
 	// Standard deviation
-	sigma float64
-	// Variance
-	variance float64
-	// Sampling precision.
-	// This parameter tells us how close we are to uniformly
-	// sampling from the interval [0,1]
-	eps float64
-	// Limit for the sampling interval.
-	// Events outside this interval are highly unlikely.
-	k float64
-
-	// The interval for sampling
-	cut int
+	sigma *big.Float
+	// Precision parameter
+	n int
+	// Precomputed values of exponential function with precision n
+	preExp []*big.Float
+	// Precomputed values for quicker sampling
+	powN *big.Int
+	powNF *big.Float
+
 }
 
 // NewNormal returns an instance of Normal sampler.
 // It assumes mean = 0.
-func NewNormal(sigma, eps, k float64) *Normal {
-	return &Normal{
+func NewNormal(sigma *big.Float, n int) *Normal {
+	powN := new(big.Int).Exp(big.NewInt(2), big.NewInt(int64(n)), nil)
+	powNF := new(big.Float)
+	powNF.SetPrec(uint(n))
+	powNF.SetInt(powN)
+	s := &Normal{
 		sigma:    sigma,
-		variance: sigma * sigma,
-		eps:      eps,
-		k:        k,
-		cut:      int(sigma * math.Sqrt(k)),
+		n:        n,
+		preExp:   nil,
+		powN:     powN,
+		powNF:    powNF,
+	}
+	return s
+}
+
+
+
+// an approximation of a exp(-x/alpha) with taylor
+// polynomial of degree k, precise at least up to 2^{-n}
+func taylorExp(x *big.Int, alpha *big.Float, k int, n int) *big.Float {
```

Comment:
This n comes from lwe schemes, where it is considered as a security parameter. Do you mean that whenever a new sampler is created it is created with uint(n)?

Reply:
Yes. If `normal`'s `n` field should never hold negative value, then it makes sense to keep it as `uint`, especially since there would otherwise be lots of conversions within its `isExpGreater` method and function `taylorExp`.

Let's leave `n` in the context of LWE schemes as is (e.g. an `int`) (it is for another PR :slightly_smiling_face: )